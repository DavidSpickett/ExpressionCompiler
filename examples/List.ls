(import "lib/lib.ls")

(let 'ls (list 1 2 3 4)
  (body
    (print "The list:" ls)
    (print "The expanded list:" *ls)

    (print "Head:" (head ls))
    (print "Last:" (last ls))
    (print "Tail:" (tail ls))
    (print "Init:" (init ls))

    (print "Apply function:")
    (defun 't 'x (print x))
    (print (apply t ls))

    (print "Apply anonymous Function:")
    (let 'a
      (apply
        (defun ' 'v
          (print v)
        )
        ls
      )
      (print a)
    )

    (let 'm
      (map
        (defun ' 'x
          (body
            (+ x 2)
          )
        )
        ls
      )
      (print "Map with + 2:" m)
    )

    (print "Accumulate:"
      # TODO: a builtin like "+" should be allowed for 'fn
      (accumulate
        (defun ' 'x 'y (+ x y))
        ls 0)
    )

    (print "Reverse:" (reverse ls))
  )
)
