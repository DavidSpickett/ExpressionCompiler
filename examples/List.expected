The list: (1, 2, 3, 4)
The expanded list: 1 2 3 4
Head: 1
Last: 4
Tail: (2, 3, 4)
Init: (1, 2, 3)
Apply function:
1
2
3
4
None
Apply anonymous Function:
1
2
3
4
None
Map with + 2: (3, 4, 5, 6)
Accumulate: 10
Reverse: (4, 3, 2, 1)
None
